addons/mail/static/src/models/dropdown_menu_view.js

```diff
@@ -0,0 +1,181 @@
+            );
+            if (activeIndex === 0) {
+                // loop when reaching the start of the list
+                this.setLastActive();
+                return;
+            }
+            const previousDropdownItemView = this.items[activeIndex - 1];
+            this.update({ userExplicitSelection: previousDropdownItemView });
+        },
+        /**
+         * Closes the popover when clicking outside, if appropriate.
+         *
+         * @private
+         * @param {MouseEvent} ev
+         */
+        _onClickCaptureGlobal(ev) {
+            if (!this.component || !this.component.root.el) {
+                return;
+            }
+            if (this.anchorRef && this.anchorRef.el && this.anchorRef.el.contains(ev.target)) {
+                return;
+            }
+            if (this.component.root.el.contains(ev.target)) {
+                return;
+            }
+            this.delete();
+        },
+    },
+    fields: {
+        activeItem: one('DropdownItemView', {
+            compute() {
+                if (this.items.includes(this.userExplicitSelection)) {
+                    return this.userExplicitSelection;
+                }
+                const firstDropdownItemView = this.items[0];
+                return firstDropdownItemView;
+            },
+        }),
+        anchorRef: attr({
+            compute() {
+                if (this.ownerAsComposerSuggestion) {
+                    return this.ownerAsComposerSuggestion.textInput.textareaRef;
+                }
+                return clear();
+            },
+            required: true,
+        }),
+        component: attr(),
+        extraItems: many('DropdownMenuExtraItemView', {
+            compute() {
+                if (this.ownerAsComposerSuggestion) {
+                    return this.ownerAsComposerSuggestion.extraSuggestions.map(suggestable => ({ suggestable }));
+                }
+                return clear();
+            },
+            inverse: 'dropdownMenuViewOwner',
+        }),
+        hasToScrollToActiveItem: attr({
+            default: false,
+        }),
+        items: many('DropdownItemView', {
+            inverse: 'dropdownMenuView',
+        }),
+        mainItems: many('DropdownMenuMainItemView', {
+            compute() {
+                if (this.ownerAsComposerSuggestion) {
+                    return this.ownerAsComposerSuggestion.mainSuggestions.map(suggestable => ({ suggestable }));
+                }
+                return clear();
+            },
+            inverse: 'dropdownMenuViewOwner',
+        }),
+        manager: one('DropdownMenuManager', {
+            compute() {
+                if (this.messaging.dropdownMenuManager) {
+                    return this.messaging.dropdownMenuManager;
+                }
+                return clear();
+            },
+            inverse: 'items',
+        }),
+        ownerAsComposerSuggestion: one('ComposerView', {
```

Comment:
name is in reverse, it should be `composerViewOwner`

The `As` part is only necessary to distinguish if there are 2 different `composerViewOwner` owners that would have a different purpose, which is not the case here.

Reply:
The naming of owner fields `modelNameOwner` is a mistake.

- It was it hard to see it's semantically and essential a "owner" field (it should comme first)
- the xor on different owners is easier to understand and see with pattern `ownerAsX`
- in `ownerAsX`, the `X` can be brief as desirable, which makes code much more readable
- the main issue with `modelNameOwner` is that it gives very long name and put too much emphasis on targeted model name, which makes whole logic rely on typing instead of actual semantic of the relational field.

Here `ownerAsComposerSuggestion` is very clear it's one contextual `owner`. `AsComposerSuggestion` makes it clear it's in the specific feature "composer suggestion".

`composerViewOwner` is far less functionally explicit